tests/integration/conftest.py

```diff
@@ -81,9 +82,48 @@ def assert_cairo_test_cases(
         skipped=set(expected_skipped_test_cases_names),
     )
 
+    name_to_test_case_result = {
+        test_result.test_case_name: test_result
+        for test_result in testing_summary.test_results
+        if isinstance(test_result, TestCaseResult)
+    }
+    diff = show_diff_between_cairo_test_cases(
+        name_to_test_case_result, expected, actual
+    )
+    if diff:
+        assert False, diff
     assert actual == expected
 
 
+def show_diff_between_cairo_test_cases(
+    name_to_test_case_result: dict[str, Any],
+    expected: CairoTestCases,
+    actual: CairoTestCases,
+) -> str:
+    lines: list[str] = []
+    result_types = ["passed", "failed", "broken", "skipped"]
+    for expected_result_type in result_types:
+        for expected_test_case_name in getattr(expected, expected_result_type):
+            if expected_test_case_name in getattr(actual, expected_result_type):
+                continue
+            lines.append(
+                (
+                    f"Expected '{expected_test_case_name}' to be {expected_result_type}, got:"
+                )
+            )
+            if expected_test_case_name not in name_to_test_case_result:
+                lines.append("nothing — no such test case result found")
+                break
+            lines.append(
+                format_test_result(
+                    name_to_test_case_result[expected_test_case_name]
+                ).format_human(LogColorProvider())
+            )
+            lines.append("")
+            break
```

Comment:
this whole flow with breaks and continues is confusing. i don't know if it's that important though since it's only used for test output